masonry/src/widgets/button.rs

```diff
@@ -14,71 +14,50 @@ use vello::peniko::Color;
 
 use crate::core::keyboard::{Key, NamedKey};
 use crate::core::{
-    AccessCtx, AccessEvent, ArcStr, BoxConstraints, ChildrenIds, EventCtx, LayoutCtx, NewWidget,
-    PaintCtx, PointerEvent, PropertiesMut, PropertiesRef, RegisterCtx, TextEvent, Update,
-    UpdateCtx, Widget, WidgetId, WidgetMut, WidgetPod,
+    AccessCtx, AccessEvent, BoxConstraints, ChildrenIds, EventCtx, LayoutCtx, NewWidget, PaintCtx,
+    PointerEvent, PropertiesMut, PropertiesRef, RegisterCtx, TextEvent, Update, UpdateCtx, Widget,
+    WidgetId, WidgetMut, WidgetPod,
 };
 use crate::properties::{
     ActiveBackground, Background, BorderColor, BorderWidth, BoxShadow, CornerRadius,
     DisabledBackground, HoveredBorderColor, Padding,
 };
 use crate::theme;
 use crate::util::{fill, include_screenshot, stroke};
-use crate::widgets::Label;
 
 /// A button with a text label.
 ///
 /// Emits [`ButtonPress`] when pressed.
 ///
 #[doc = include_screenshot!("button_hello.png", "Button with text label.")]
 pub struct Button {
-    label: WidgetPod<Label>,
+    child: WidgetPod<dyn Widget>,
 }
 
 // --- MARK: BUILDERS
 impl Button {
-    /// Create a new button with a text label.
+    /// Create a new button with a child widget.
     ///
     /// # Examples
     ///
     /// ```
-    /// use masonry::widgets::Button;
-    ///
-    /// let button = Button::new("Increment");
-    /// ```
-    pub fn new(text: impl Into<ArcStr>) -> Self {
-        Self::from_label(Label::new(text).with_auto_id())
-    }
-
-    /// Create a new button with the provided [`Label`].
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use masonry::core::{StyleProperty, Widget as _};
-    /// use masonry::peniko::Color;
     /// use masonry::widgets::{Button, Label};
+    /// use masonry::core::Widget;
     ///
-    /// let label = Label::new("Increment").with_style(StyleProperty::FontSize(20.0));
-    /// let button = Button::from_label(label.with_auto_id());
+    /// let button = Button::new(Label::new("Increment").with_auto_id());
     /// ```
-    pub fn from_label(label: NewWidget<Label>) -> Self {
+    pub fn new(child: NewWidget<impl Widget>) -> Self {
         Self {
-            label: label.to_pod(),
+            child: child.erased().to_pod(),
         }
     }
 }
 
 // --- MARK: WIDGETMUT
 impl Button {
-    /// Set the text.
-    pub fn set_text(this: &mut WidgetMut<'_, Self>, new_text: impl Into<ArcStr>) {
-        Label::set_text(&mut Self::label_mut(this), new_text);
-    }
```

Comment:
I wish there was a way to keep that method. I don't see an obvious one, though.

Reply:
This might be a little hard as we will need to make assumptions of the child widget type. We could do `try_downcast` and return a `Result` or a `bool` but I feel like it's not ergonomic to use.